### DISCUSSION TITLE
Can IPAM be used to manage AVS devices

### DISCUSSION THREAD

**Discussion by tonydean37** [Azure/ipam#81] [General]
Hi,

This project looks really interesting and sounds ideal for keeping track of native Azure IP addresses. I was wondering if it has the ability to read the information in an AVS instance so we can automatically track/monitor usage in that environment?

Thanks

**Comment by DCMattyG** [comment]
Hey @tonydean37, I'll make the assumption you're referring to Azure VMware Solution here. At this moment, Azure IPAM isn't aware of the address space consumed in the Private Clouds within an AVS. That said, I'll get working on integrating it right away, since I agree visibility into AVS Private Clouds would be extremely useful here.

Thanks for the suggestion!

**Reply by tonydean37** [reply]
Hi again @DCMattyG,

Perfect - Azure VMware Solution is exactly the one I was looking at. :) In terms of being able to populate IPs, do you think it will be possible to obtain both the NSX-T information as well as the VMs themselves? I understand you may have to do some digging before being able to answer this one though!

**Reply by tonydean37** [reply]
@DCMattyG I have another couple of queries on the services that Azure IPAM can manage, if I may:
1. Is it possible to discover the external IPs that are currently assigned to devices/services?
2. If you have an Azure secured vWAN Hub, is it possible to discover the IPs of the devices within this (e.g. the Azure firewall, ExpressRoute IPs etc...)

**Reply by DCMattyG** [reply]
Hi @tonydean37, answers as follows:

1. At the moment, no. This tool was designed primarily to focus on things within the Azure space, although we've discussed something like the ability to add "static" entries to represent things outside of Azure. Is there anything specific you had in mind?
2. Funny you should ask this, I'm actually working on vWAN Hubs as we speak. This functionality is almost complete.

**Reply by DCMattyG** [reply]
BTW, at first glance it does indeed look like we can retrieve both NSX-T and VM info from the current AVS API's. Time to do some testing!

**Reply by tonydean37** [reply]
Thanks for this - it'll be great having AVS, NSX-T and vWAN support. I'm looking forward to giving this a test drive as we currently use a spreadsheet for our IPAM! Once I have the PoC setup then please let me know if there is anything I can help in regards to testing.

The addition of being able to add static entries would be brilliant, this would allow us to add our on-prem devices into the solution too, which would mean a single pane of glass for our whole environment.

Another query - sorry for all the questions! Does the solution have the ability to discover and manage Public IPs in Azure? For example, we have around 200 PIPs associated to various device types (NAT Gateways, Load balancers, Application Gateways and some that are unassociated etc...) and it would be useful to be able to have those managed as part of this.

**Reply by DCMattyG** [reply]
Hi @tonydean37 , first and foremost please ask as many questions as you'd like. We're here to help you be successful and to make sure everyone enjoys our IPAM project. 

It's funny you should mention Public IP's as I was just having this conversation with a colleague the other day...

I'm curious what you mean by "managed" in the context of Azure Public IP's. Generally these IP's are pulled from Azure-owned blocks and not in the control of the customer, and often are not able to be summarized into a large CIDR as they can be highly non-contiguous. We're you just looking for a way to track the currently assigned Public IP's for various endpoints, or are you implying something else?

**Reply by tonydean37** [reply]
Brill - thank you. :)

The major benefit would be to be able to discover these, so we have a record of them in a single IPAM solution. Ideally including the DNS name and associated device where possible. 

When I saw that we had about 30 that are not associated with anything, it made me think if we could somehow integrate this into automation? So if a user needed a public IP, check to see if there is one that isn't associated to anything and use that. I think that could become a bit complicated though and could cause issues further down the line if someone has created a PIP for their own use and not assigned it to anything yet, so the more I think about it, then the less I think it is required.

As you say, the fact that PIPs are highly non-contiguous, I'm not too sure how this would look in the database, but it would be nice if these were all in a single section, to provide a quick overview of PIP usage.

**Reply by tonydean37** [reply]
Morning @DCMattyG, I've been having a bit of a play with Azure Resource Graph and seeing what I can pull out regarding IPs. I saw a workbook Clive Watson had written, so have added a bit of code, so we can pull out the PIP of the Azure firewalls we have too. I can then create a workbook for PIPs so we have a record somewhere.

Wondered if you thought this would cover all PIPs, or if there are other resource types I would need to look at too?

```
Resources
//| where (type == "microsoft.network/publicipaddresses" or type == "microsoft.network/azurefirewalls")
| extend Details = pack_all()
| extend FWipAddress = extract(@'"address":"(.*)"',1,tostring(properties.hubIPAddresses.publicIPs))
| extend DNSName = extract(@'domainNameLabel":(.*),',1,tostring(properties.dnsSettings))
| extend FQDN = extract(@'fqdn":(.*)',1,tostring(properties.dnsSettings))
| extend AttachedResource = extract(@':"(.*)"',1,tostring(properties.ipConfiguration))
| extend IPAddress = strcat(FWipAddress, tostring(properties.ipAddress))
| summarize count() by subscriptionId, resourceGroup, type
, publicIPAllocationMethod = tostring(properties.publicIPAllocationMethod), name, IPAddress, FWipAddress, DNSName, FQDN
, tostring(properties.ipAddress)
, tostring(properties.provisioningState)
, tostring(properties.publicIPAddressVersion)
, AttachedResource
| where isnotempty(publicIPAllocationMethod) or isnotempty(FWipAddress)
| project subscriptionId, resourceGroup, ['type'], name, IPAddress, publicIPAllocationMethod, DNSName, FQDN, properties_provisioningState, properties_publicIPAddressVersion, AttachedResource
| order by IPAddress desc
```

**Reply by DCMattyG** [reply]
Thank you for this @tonydean37! I'll definitely use this as a starting point. My intention is to have a separate page where users can view all of the Public IP information, and it will also have the PIP's matched to their attached resources where applicable.

I'm wrapping up the vWan Hub work today, and then starting the AVS exploration next. Public IP's are in my queue right after we see what we can do for AVS.

**Reply by tonydean37** [reply]
This all sounds really great - thank you :). Appreciate the speedy responses to all the queries.

Have a great weekend!
